src/main/java/net/neoforged/neoforge/transfer/resources/IResourceStack.java

```diff
@@ -0,0 +1,146 @@
+     *     },
+     *     "amount": 1000
+     * }
+     * }</pre>
+     *
+     * @param resourceCodec a codec for the resource
+     * @param <R>           the resource type
+     * @return a codec for a resource stack
+     */
+    static <R extends IResource, S extends IResourceStack<R>> Codec<S> codec(Codec<R> resourceCodec, BiFunction<R, Integer, S> factory) {
+        return RecordCodecBuilder.create(instance -> instance.group(
+                resourceCodec.fieldOf("resource").forGetter(IResourceStack<R>::resource),
+                NeoForgeExtraCodecs.optionalFieldAlwaysWrite(ExtraCodecs.NON_NEGATIVE_INT, "amount", 1)
+                        .forGetter(IResourceStack<R>::amount))
+                .apply(instance, factory));
+    }
+
+    /**
+     * Creates a codec where the fields for the resource are at the same level as the amount
+     *
+     * <pre>{@code
+     * {
+     *    "id": "minecraft:water",
+     *    "components": { ... },
+     *    "amount": 1000
+     * }
+     * }</pre>
+     *
+     * @param resourceCodec Backing resource codec
+     * @param factory       Constructor of IResourceStack implementation
+     * @param <R>           The resource type
+     * @return Codec for the specified IResourceStack implementer
+     */
+    static <R extends IResource, S extends IResourceStack<R>> Codec<S> flatCodec(Codec<R> resourceCodec, BiFunction<R, Integer, S> factory) {
+        return RecordCodecBuilder.create(instance -> instance.group(
+                MapCodec.assumeMapUnsafe(resourceCodec).forGetter(IResourceStack<R>::resource),
+                NeoForgeExtraCodecs.optionalFieldAlwaysWrite(ExtraCodecs.NON_NEGATIVE_INT, "amount", 1)
+                        .forGetter(IResourceStack<R>::amount))
+                .apply(instance, factory));
+    }
+
+    /**
+     * Creates a standard stream codec for a IResourceStack implementer of the specified resource type.
+     */
+    static <B extends FriendlyByteBuf, R extends IResource, S extends IResourceStack<R>> StreamCodec<B, S> streamCodec(StreamCodec<? super B, R> resourceCodec, BiFunction<R, Integer, S> factory) {
+        return StreamCodec.composite(
+                resourceCodec, IResourceStack::resource,
+                ByteBufCodecs.VAR_INT, IResourceStack::amount,
+                factory);
+    }
+
+    /**
+     * @return the backing resource of the stack.
+     */
+    T resource();
+
+    /**
+     * @return the amount currently set in the stack
+     */
+    @Nonnegative
+    int amount();
+
+    /**
+     * Checks if this is empty, meaning that the amount is not positive
+     * or that the resource is {@link IResource#isEmpty() empty}.
+     *
+     * @return {@code true} if empty
+     */
+    default boolean isEmpty() {
+        return amount() <= 0 || resource().isEmpty();
+    }
+
+    default boolean isEnabled(FeatureFlagSet enabledFeatures) {
+        return resource() instanceof FeatureElement element && element.isEnabled(enabledFeatures);
```

Comment:
Couple things wrong here that i just noticed
- None of the current `IResource` interfaces/implementations extend/implement `FeatureElement`.
  - Due to this resources would be assumed to always be disabled
  - There is however `IRegisteredResource.isEnabled`, might want to change to invoke this method instead of `FeatureElement.isEnabled`.
- The boolean logic here is incorrect. Currently any none `FeatureElement` resources would be treated as always being disabled and only `FeatureElement` resources could be enabled.
  - The logic here should ask any `FeatureElement` resources if they should be enabled, any others should be assumed to be enabled.

```suggestion
        return !(resource() instanceof IRegisteredResource reg) || reg.isEnabled(enabledFeatures);
```

Reply:
For the first note, should we make the IRegisteredResource implement FeatureElement?